### ISSUE TITLE
Request for additional prompt elements

### ISSUE THREAD

**Issue by z0rc** [romkatv/powerlevel10k#342]
Would it be possible to add elements to indicate status from:
* https://github.com/tokuhirom/plenv
* https://github.com/cehoffman/luaenv
* https://github.com/syndbg/goenv
* https://github.com/jenv/jenv

All in the same approach like pyenv and rbenv. After all they are similar in implementation, just for different languages.

Also in addition to `ranger` file manager element, it would be nice to have indicators for [`mc`](https://github.com/MidnightCommander/mc) (using `MC_SID` env var) and [`nnn`](https://github.com/jarun/nnn) (using `NNNLVL` env var). Maybe make a generic file manager element as it's very unlikely the user would be using multiple nested filemanagers simultaneously.

**Comment by romkatv** [comment]
> Would it be possible...

Yes. It just requires time.

Which of the listed prompt segments do you need?

**Comment by z0rc** [comment]
> Which of the listed prompt segments do you need?

Technically each of them. But most useful right now would be mc and goenv.

**Comment by romkatv** [comment]
> > Which of the listed prompt segments do you need?
> 
> Technically each of them. But most useful right now would be mc and goenv.

I've implemented `midnight_commander` and `goenv` prompt segments. Both are listed in config templates used by `p10k configure`. `goenv` is enabled by default, `midnight_commander` isn't.

**Comment by extremepayne** [comment]
Alright. I'd like to contribute to p10k since it's so awesome and looks a bit like a one-man-show that's about to hit some growing pains. This looks like a good place to start as any. 
I looked at the commit for goenv and it doesn't look too hard, I'd have to brush up on bash/zsh syntax to really grok it all, but I could probably get a patch out pretty quick for one of the others (like luaenv). Is there anything I should be aware of before starting? It doesn't look like you have a test suite so that shouldn't be a problem, and the changes look fairly simple so should be quick to replicate. A couple of questions:
What is going on on line 5337/5379 of internal/p10k.zsh ([on the goenv commit](https://github.com/romkatv/powerlevel10k/commit/ac2f387dfd68007801ab75004b076e89ce88eab9))? You just increment one number, why?
Are there any benchmarks I should run before submitting a patch?
And finally, any special things I should be aware of? You've mentioned rewriting C APIs for a couple of the other segments (pyenv, node, I think). I didn't see this happening here; do you anticipate that would be necessary for any of the above requests?

**Comment by romkatv** [comment]
> a one-man-show that's about to hit some growing pains

There are 0 open PRs and 11 open issues, one of which is a minor known bug that I don't know how to fix, two are medium impact features, and the rest are low impact feature requests.requests.

There aren't many projects with comparable number of stars/users with such low number of open issues and PRs.

> Is there anything I should be aware of before starting?

The hard part is understanding the logic of the underlying project (e.g., `luaenv`) so that you can reimplement it and be reasonably confident that your reimplementation works the same way. This can be time consuming. In the worst case it takes me a full day for a single `*env` (have to read and trace through thousands of lines of bad bash). In the best case it's about an hour. The best case happens when all of the following requirements are met:

- The new `*env` is a fork of another `*env` for which I've already implemented a prompt segment. (Note: `luaenv` is a fork of `rbenv`.)
- The fork hasn't diverged much. (Note: `luaenv` has only a few commits on top of `rbenv`.)

In cases like this I can go through the diff against the parent project.

> A couple of questions:
> What is going on on line 5337/5379 of internal/p10k.zsh ([on the goenv commit](https://github.com/romkatv/powerlevel10k/commit/ac2f387dfd68007801ab75004b076e89ce88eab9))? You just increment one number, why?

This is necessary whenever a new parameter is added. In this case it was `POWERLEVEL9K_GOENV_PROMPT_ALWAYS_SHOW`. It's OK to bump the version number even when it's not necessary. However, if you forget to bump it when it *is* necessary, the newly added parameter may end up having incorrect default value.

> Are there any benchmarks I should run before submitting a patch?

No.

> And finally, any special things I should be aware of?

`luaenv` looks easy enough to support. On one hand this is good as you should be able to do it (unless there are hidden gotchas in the diff against `rbenv`.) On the other hand it means you won't be saving me any time. The effort to verify the correctness of your code will be the same as the effort to implement `luaenv` segment from scratch because the hard part is to ensure that copy-pasting `rbenv` prompt segment and replacing ruby with lua is safe.

**Comment by extremepayne** [comment]
Alright. I'm willing to do contributions anytime you need them, but it sounds like I'd do more harm than help here. And you're right, the project is moving _really_ fast, I was just referring to the fact that many feature requests have been deferred with "I don't have much time."

**Comment by romkatv** [comment]
> "I don't have much time."

It's more accurate to say that the impact of these features (in my own view) is too low to justify time investment. In addition, I have less time nowadays that a month ago because of holidays.

**Comment by romkatv** [comment]
This is done.
